profiling/src/exporter/mod.rs

```diff
@@ -120,40 +124,61 @@ impl ProfileExporter {
     }
 
     /// Build a Request object representing the profile information provided.
+    #[allow(clippy::too_many_arguments)]
     pub fn build(
         &self,
         start: DateTime<Utc>,
         end: DateTime<Utc>,
         files: &[File],
         additional_tags: Option<&Vec<Tag>>,
         timeout: std::time::Duration,
+        profiling_library_name: &str,
+        profiling_library_version: &str,
     ) -> anyhow::Result<Request> {
         let mut form = multipart::Form::default();
 
-        form.add_text("version", "3");
-        form.add_text("start", start.format("%Y-%m-%dT%H:%M:%S%.9fZ").to_string());
-        form.add_text("end", end.format("%Y-%m-%dT%H:%M:%S%.9fZ").to_string());
-        form.add_text("family", self.family.as_ref());
-
-        for tags in self.tags.as_ref().iter().chain(additional_tags.iter()) {
-            for tag in tags.iter() {
-                form.add_text("tags[]", tag.to_string());
-            }
+        // combine tags and additional_tags
+        let mut tags_profiler = String::new();
+        let other_tags = additional_tags.into_iter();
+        for tag in self.tags.iter().chain(other_tags).flatten() {
+            tags_profiler.push_str(tag.as_ref());
+            tags_profiler.push(',');
         }
+        tags_profiler.pop(); // clean up the trailing comma
+
+        let attachments: Vec<String> = files.iter().map(|file| file.name.to_owned()).collect();
+
+        let event = json!({
+            "attachments": attachments,
+            "tags_profiler": tags_profiler,
+            "start": start.format("%Y-%m-%dT%H:%M:%S%.9fZ").to_string(),
+            "end": end.format("%Y-%m-%dT%H:%M:%S%.9fZ").to_string(),
+            "family": self.family.as_ref(),
+            "version": "4",
+        })
+        .to_string();
+
+        // event.json shouldn't be compressed
+        form.add_reader_file_with_mime(
+            "event",
+            Cursor::new(event),
+            "event.json",
+            mime::APPLICATION_JSON,
+        );
 
         for file in files {
-            form.add_reader_file(
-                format!("data[{}]", file.name),
-                Cursor::new(file.bytes.to_owned()),
-                file.name,
-            )
+            let mut encoder = FrameEncoder::new(Vec::new());
+            encoder.write_all(file.bytes)?;
+            form.add_reader_file(file.name, Cursor::new(encoder.finish()?), file.name)
```

Comment:
How is the compression format defined in the payload ?

Reply:
AFAIK the intake just looks for magic bits to distinguish if an uploaded file is [gzip|lz4] compressed or not. Otherwise no changes needed.

Reply:
Ivo is right, according to Florian:

> Levi: How does it know what compression format to use? Inspect magic bytes?
> Florian: yes
> Florian: I was looking at java profiler and it uses LZ4 by default
> Florian: Jaroslav experiments has shown that it's a good space/CPU tradeoff for the profiler

It's also is why I picked lz4. I didn't test or verify these claims beyond "works for me."

Reply:
Thanks. There are indeed magic values at the start of the payload that you can use. 
As a Datadog agnostic library, this is not super friendly.